feat(devtools): add support for @for control flow blocks in directive explorer

Avcharov opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?
- [x] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

@for loops arent present in devtools tree

Issue Number: #65849 

## What is the new behavior?

https://github.com/user-attachments/assets/5d8e5115-2739-4ce8-86b5-7276d7c10336

- Handle @for data in tree strategies and view extraction
- Show @for details in the UI and property tab
- Persist @for state and update UI accordingly

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

JeanMeche commented on packages/core/src/render3/util/repeater.ts:
```diff
@@ -0,0 +1,129 @@
+ * Gets all of the `@for` loop blocks that are present inside the specified DOM node.
+ * @param node Node in which to look for `@for` blocks.
+ *
+ * @publicApi
```
Since this is exposed with a private symbol `ɵForLoopBlockData`, this isn't a public api entry. 
```suggestion
```

Avcharov committed:
feat(devtools): add support for @for control flow blocks in directive explorer

- Handle @for data in tree strategies and view extraction
- Show @for details in the UI and property tab
- Persist @for state and update UI accordingly

dgp1130 commented:
Thanks for putting this together! I can definitely see value in inspecting the model in the `@for` independent of what got rendered from it. I'll defer to @hawkgs and @AleksanderBodurri on a proper review. The two general points of feedback I have to consider are:

1. I think we've discussed consolidating some of the `@` blocks into a more generic structure in the `ng` global, such that we could combine `@defer`, `@for`, `@if`, and `@switch` under the same mechanism and avoiding the need to create custom accessors for each kind of block, as well as make it easier to support new blocks in the future. Something like `ng.getBlocks(ng.getComponent($0))` and returning some Abstract Data Type enumerating all the types of blocks with internal data (what's showed in the property explorer) and child nodes. I know some Wiz people are starting to look at replacing DOM walking by asking the application on the page via the `ng` global what the component tree looks like and we could potentially carve out some API space there for non-component blocks like `@for`, though I'm not sure what the current status of that effort is.

2. UX note: Would there be value in showing the actual `track` value for each item in the list? I see it shows the `track` function, which is great to include as well, but I suspect it would be more useful to see the actual runtime `track` value for each item to be able to know if it matches what the developer expects. Relatedly, I wonder if there might be value in highlighting duplicated `track` values across multiple items, though I'm not sure if we have other warnings for that scenario. We can also consider this in a future PR, doesn't necessarily need to be blocking here.

hawkgs commented on devtools/projects/ng-devtools-backend/src/lib/directive-forest/render-tree.ts:
```diff
@@ -71,103 +89,154 @@ const extractViewTree = (
-    );
+    if (nodesToSkip.has(node)) {
+      return;
+    }
```
Question: Is the reason for moving this in the beginning of the block (in combination with group functions returning a boolean), intended to prevent calling the group functions with the control flow children that are already marked for skipping?

hawkgs commented on devtools/projects/ng-devtools-backend/src/lib/directive-forest/render-tree.ts:
```diff
@@ -71,103 +89,154 @@ const extractViewTree = (
+    },
+  });
+
+  currentDeferBlock.rootNodes.forEach((child) => nodesToSkip.add(child));
```
I know it's part of an old, but could you convert `forEach`-es to `for ... of` in the affected LoC.

hawkgs commented on devtools/projects/ng-devtools-backend/src/lib/state-serializer/state-serializer.ts:
```diff
@@ -129,8 +129,8 @@ export function serializeDirectiveState(instance: object): Record<string, Descri
-  return levelSerializer({value}, 'value', false, 0, 0);
+export function serializeValue(value: unknown, level = MAX_LEVEL): Descriptor {
+  return levelSerializer({value}, 'value', false, 0, level);
 }
```
I suspect that this change will cause a bug. `MAX_LEVEL` is `1` which changes the default level from `0` to `1`.

hawkgs commented on devtools/projects/ng-devtools-backend/src/lib/directive-forest/render-tree.ts:
```diff
@@ -71,103 +89,154 @@ const extractViewTree = (
+  children: ComponentTreeNode[],
+  ctx: TreeExtractionContext,
+): ComponentTreeNode {
+  const serializedItems = forLoop.items.map((item) => serializeValue(item, 5));
```
Why is the level set to `5`?

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/directive-explorer/directive-explorer.component.ts:
```diff
@@ -205,6 +205,17 @@ export class DirectiveExplorerComponent {
+          this._clickedElement = updatedNode;
+        }
+      }
+
```
Same comment regarding relying on the position.

hawkgs commented on packages/core/src/render3/util/repeater.ts:
```diff
@@ -0,0 +1,139 @@
+export interface ForLoopBlockData {
+  items: unknown[];
+  hasEmptyBlock: boolean;
+  itemCount: number;
```
Do we need the `itemCount`? It feels that this can be inferred from `items`.

hawkgs commented on packages/core/src/render3/util/repeater.ts:
```diff
@@ -0,0 +1,139 @@
+  }
+
+  const fnForDisplay = metadata.originalTrackByFn ?? trackByFn;
+  return fnForDisplay.toString();
```
I am not really sure we would want to stringify the track by function. I'd rather print just `function`.

hawkgs commented on packages/core/src/render3/util/repeater.ts:
I guess it would be good to have a second pair of eyes on this one.

cc @JeanMeche , @crisbeto

hawkgs commented:
JFYI: I took over the PR. We will abstract the current solution so we can easily add support for the upcoming control flow blocks.

Work in progress.

Avcharov commented:
Got you, sorry for not appearing, quite busy for now.
If you need some additional comment from my side for your not resolved comments - let me know, i will respond ASAP

Sorry one more time for not responding before

hawkgs committed:
refactor(core): abstract control flow discovery utilities

Create a universal mechanism for adding and managing control flow discovery utilities. Those utilities are intended to be consumed by third parties via the `ng` global interface.

hawkgs committed:
refactor(devtools): utilize new unified getControlFlowBlocks API

Utilize the new `ng.getControlFlowBlocks` API by refactoring the existing render tree extraction code; Slighly modify @for loop details view

hawkgs commented:
The PR is ready for a review. In addition to the original commit (`2ddb750`) that introduces the `@for` loop block support, the two additional commits that target separately `core` and `devtools`, are substituting the existing `getDeferBlocks` and the new `getForLoopBlocks` in favor of a universal `getControlFlowBlocks` that we can utilize for future extended control flow support as well. As already noted, the changes affect mainly the framework part in charge of `@defer` and `@for` discovery (2nd commit) and the existing render tree extraction logic in the DevTools (3rd commit).

cc @dgp1130, @JeanMeche

dgp1130 commented on packages/core/src/render3/util/control_flow.ts:
```diff
@@ -0,0 +1,350 @@
+ * @param lView View within the node in which to search for blocks.
+ * @param results Array to which to add blocks once they're found.
+ */
+function findControlFlowBlocks(node: Node, lView: LView, results: ControlFlowBlock[]) {
```
Nit: Use explicit return type.

dgp1130 commented on packages/core/src/render3/util/control_flow_types.ts:
```diff
@@ -0,0 +1,108 @@
+ */
+export interface ControlFlowBlockViewFinder {
+  (config: ControlFlowBlockViewFinderConfig): ControlFlowBlock | null;
+}
```
Consider: It may be more readable to express this as a type alias to a function rather than a callable interface.

```suggestion
export type ControlFlowBlockViewFinder = (config: ControlFlowBlockViewFinderConfig) => ControlFlowBlock | null;
```

dgp1130 reviewed (approved):
Sorry for the delay, I gave some general suggestions and don't have any major objections, but TBH most of the `LView` stuff goes over my head. I'm hoping @JeanMeche is more familiar with the intricacies here or else we can pull in someone from the framework team with more knowledge about how reasonable this usage is.

I only really looked at the second two commits, should we squash in whatever we're using/retaining from the first commit?

hawkgs commented on packages/core/src/render3/util/control_flow.ts:
```diff
@@ -0,0 +1,350 @@
+  if (trackByFn.name === 'ɵɵrepeaterTrackByIdentity') {
+    return 'item';
+  }
+  return 'function';
```
Greg did something similar in the initial commit by stringifying the source function. 

Are you fine deferring this to another PR since it'll require modifying the Ivy control flow instructions file – ideally having a dev-mode-only code to avoid unnecessary overhead – so it will involve/demand a more careful change?

I'll add it to my backlog.

hawkgs committed:
fixup! refactor(core): abstract control flow discovery utilities

JeanMeche reviewed (approved):
🚢 it!
